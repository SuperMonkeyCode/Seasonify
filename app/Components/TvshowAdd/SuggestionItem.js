/* @flow */
import {
  Body,
  Button,
  Icon,
  Left,
  ListItem,
  Right,
  Text,
  Thumbnail,
  View,
} from 'native-base';
import { Keyboard } from 'react-native';
import { compose, pure, withHandlers } from 'recompose';
import React from 'react';

type Props = {
  /* parent */
  navigation: Object,
  suggestionKey: number,
  alreadyAddedId?: string,
  onPress: Function,
  poster?: string,
  title: string,
  subtitle?: string,
  /* connect */
  /* HOC */
  handlePressSuggestion: Function,
};

const enhance = compose(
  withHandlers({
    handlePressSuggestion: ({
      navigation,
      suggestionKey,
      alreadyAddedId,
      onPress,
    }: Props) => () => {
      if (alreadyAddedId) {
        Keyboard.dismiss();
        navigation.navigate('TvshowDetailsPage', { tvshowId: alreadyAddedId });
      } else {
        onPress(suggestionKey);
      }
    },
  }),
  pure,
);

function SuggestionItem({
  poster,
  title,
  subtitle,
  alreadyAddedId,
  handlePressSuggestion,
}: Props) {
  return (
    <View>
      <ListItem avatar onPress={handlePressSuggestion}>
        <Left>
          {poster && <Thumbnail square source={{ uri: poster }} />}
          {!poster &&
            <Button transparent disabled style={{ width: 57 }} /> /* FIXME: too tricky */}
        </Left>
        <Body>
          <Text>
            {title}
          </Text>
          <Text note>
            {subtitle}
          </Text>
        </Body>
        {!alreadyAddedId &&
          <Right>
            <Icon name="add-circle" />
          </Right>}
      </ListItem>
    </View>
  );
}

export default enhance(SuggestionItem);
