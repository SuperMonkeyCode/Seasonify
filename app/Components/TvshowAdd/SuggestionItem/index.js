/* @flow */
export { default } from './SuggestionItem';
