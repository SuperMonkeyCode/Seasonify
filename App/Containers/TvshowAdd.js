import React from 'react';
import { connect } from 'react-redux';
import { Keyboard } from 'react-native';
import {
  Container,
  Content,
  Form,
  Label,
  Input,
  Item,
  List,
} from 'native-base';
import debounce from 'throttle-debounce/debounce';

import { tvshowActions } from '../Redux/tvshowRedux';
import { uiActions, uiSelectors } from '../Redux/uiRedux';
import HeaderModular from '../Components/HeaderModular';
import SuggestionItem from '../Components/SuggestionItem';

const mapStateToProps = state => ({
  suggestions: uiSelectors.getSuggestions(state),
});

const mapActionsToProps = {
  addTvshowWithSeasons: tvshowActions.addTvshowWithSeasons,
  suggestionsRequest: uiActions.suggestionsRequest,
};

function TvshowAdd({
  navigation,
  suggestions,
  addTvshowWithSeasons,
  suggestionsRequest,
}) {
  const onChangeName = debounce(500, name => {
    suggestionsRequest(name);
  });

  const onEndName = event => {
    const name = event.nativeEvent.text;
  };

  const onPressSuggestion = suggestionKey => {
    addTvshowWithSeasons(suggestions[suggestionKey]);
    navigation.goBack();
    Keyboard.dismiss();
  };

  return (
    <Container>
      <Content>
        <Form>
          <Item fixedLabel>
            <Label>Tvshow&rsquo;s name:</Label>
            <Input
              onChangeText={onChangeName}
              onEndEditing={onEndName}
              autoFocus
              autoCapitalize="words"
            />
          </Item>
        </Form>
        {suggestions &&
          <List>
            {suggestions.map((suggestion, key) => (
              <SuggestionItem
                key={suggestion.allocine}
                suggestionKey={key}
                onPress={onPressSuggestion}
                poster={suggestion.poster}
                title={suggestion.name}
                subtitle={suggestion.year}
              />
            ))}
          </List>}
      </Content>
    </Container>
  );
}

TvshowAdd.navigationOptions = ({ navigation }) => ({
  header: (
    <HeaderModular
      title="New TV tvshow"
      cancelButton={{ icon: 'close', action: navigation.goBack }}
    />
  ),
});

export default connect(mapStateToProps, mapActionsToProps)(TvshowAdd);
