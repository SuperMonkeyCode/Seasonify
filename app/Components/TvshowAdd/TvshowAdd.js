/* @flow */
import { Container, Content, Form, Label, Input, Item, List } from 'native-base';
import { Keyboard } from 'react-native';
import { compose, pure, withHandlers } from 'recompose';
import React from 'react';
import debounce from 'throttle-debounce/debounce';

import SuggestionItem from './SuggestionItem';

type Props = {
  /* parent */
  navigation: Object,
  /* connect */
  suggestions: Tvshow[],
  codes: { [id: string]: number },
  tvshowAddWithSeasons: Function,
  suggestionsRequest: Function,
  /* HOC */
  registerInput: Function,
  handleChangeName: Function,
  handlePressSuggestion: Function,
};

const enhance = compose(
  pure,
  withHandlers(
    (/* props */) => {
      let input;
      return {
        registerInput: () => ref => {
          input = ref;
        },
        handleChangeName: ({ suggestionsRequest }: Props) =>
          debounce(500, (name: string) => {
            suggestionsRequest(name);
          }),
        handlePressSuggestion: ({ suggestions, tvshowAddWithSeasons }: Props) => (
          suggestionKey: number,
        ) => {
          tvshowAddWithSeasons(suggestions[suggestionKey]);
          /* istanbul ignore next */
          if (input) {
            input.wrappedInstance.clear();
            Keyboard.dismiss();
          }
        },
      };
    },
  ),
);

function TvshowAdd({
  suggestions,
  registerInput,
  codes,
  navigation,
  handleChangeName,
  handlePressSuggestion,
}: Props) {
  suggestions.forEach(suggestion => {
    suggestion.alreadyAddedId =
      Object.keys(codes).find(id => codes[id] === suggestion.allocine) || undefined;
  });

  return (
    <Container>
      <Content keyboardShouldPersistTaps="handled" keyboardDismissMode="none">
        <Form>
          <Item fixedLabel>
            <Label>Tvshow&rsquo;s name:</Label>
            <Input
              onChangeText={handleChangeName}
              autoFocus
              autoCapitalize="words"
              ref={/* istanbul ignore next */ ref => registerInput(ref)}
            />
          </Item>
        </Form>
        {suggestions &&
          <List>
            {suggestions.map((suggestion, key) =>
              (<SuggestionItem
                key={suggestion.allocine}
                navigation={navigation}
                suggestionKey={key}
                alreadyAddedId={suggestion.alreadyAddedId}
                onPress={handlePressSuggestion}
                poster={suggestion.poster}
                title={suggestion.name}
                subtitle={suggestion.year.toString()}
              />),
            )}
          </List>}
      </Content>
    </Container>
  );
}

export default enhance(TvshowAdd);
