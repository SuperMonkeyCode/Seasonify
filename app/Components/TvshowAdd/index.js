/* @flow */
import { connect } from 'react-redux';
import React from 'react';

import { tvshowActions, tvshowSelectors } from 'Store/tvshowStore';
import { uiActions, uiSelectors } from 'Store/uiStore';
import HeaderModular from 'Components/HeaderModular';

import TvshowAdd from './TvshowAdd';

TvshowAdd.navigationOptions = ({ navigation }) => ({
  header: (
    <HeaderModular
      title="New TV Show"
      cancelButton={{ icon: 'arrow-back', action: navigation.goBack }}
    />
  ),
});

export default connect(
  state => ({
    suggestions: uiSelectors.getSuggestions(state),
    codes: tvshowSelectors.getCodes(state),
  }),
  {
    tvshowAddWithSeasons: tvshowActions.tvshowAddWithSeasons,
    suggestionsRequest: uiActions.suggestionsRequest,
  },
)(TvshowAdd);
