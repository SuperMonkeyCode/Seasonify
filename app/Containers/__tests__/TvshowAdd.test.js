import React from 'react';

import { TvshowAdd } from '../TvshowAdd';

function setup(specificProps = {}) {
  const props = {
    navigation: {
      navigate: jest.fn(),
      goBack: jest.fn(),
      state: { params: {} },
    },
    codes: [111, 222],
    suggestions: [
      {
        allocine: 111,
        name: 'A tvshow',
        year: 2011,
        poster: 'http://poster1.png',
      },
      {
        allocine: 555,
        name: 'A other tvshow',
        year: 2015,
        poster: 'http://poster5.png',
      },
    ],
    tvshowAddWithSeasons: jest.fn(),
    suggestionsRequest: jest.fn(),
    ...specificProps,
  };
  const component = shallow(<TvshowAdd {...props} />);
  return {
    component,
    props,
  };
}

describe('Rendering, with 2 suggestions, one being already added', () => {
  const { component } = setup();
  it('should match', () => {
    expect(component).toMatchSnapshot();
  });
});

describe('Rendering, with no suggestions', () => {
  const { component } = setup({ suggestions: [] });
  it('should match', () => {
    expect(component).toMatchSnapshot();
  });
});

/* ========= Events & Functions ========= */

describe('Events & Functions, with 2 suggestions', () => {
  const { component, props } = setup();
  const Input = component.find('Styled(Input)');
  const SuggestionItem = component.find('SuggestionItem').first();

  describe('when calling onChangeText', () => {
    jest.useFakeTimers(); // because function is debounced
    Input.prop('onChangeText')('New name');

    it('calls suggestionsRequest', () => {
      jest.runAllTimers(); // because function is debounced
      expect(props.suggestionsRequest).toBeCalledWith('New name');
    });
  });

  describe('when calling onPress', () => {
    SuggestionItem.prop('onPress')(0);

    it('calls tvshowAddWithSeasons', () => {
      expect(props.tvshowAddWithSeasons).toBeCalledWith(props.suggestions[0]);
    });
  });
});
