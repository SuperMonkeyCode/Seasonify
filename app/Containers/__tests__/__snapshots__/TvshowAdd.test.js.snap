// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering, with 2 suggestions, one being already added should match 1`] = `
<Styled(Container)>
  <Styled(Content)>
    <Styled(Form)>
      <Styled(Item)
        fixedLabel={true}
      >
        <Styled(Label)>
          Tvshow’s name:
        </Styled(Label)>
        <Styled(Input)
          autoCapitalize="words"
          autoFocus={true}
          onChangeText={[Function]}
        />
      </Styled(Item)>
    </Styled(Form)>
    <Styled(List)>
      <SuggestionItem
        alreadyAdded={true}
        onPress={[Function]}
        poster="http://poster1.png"
        subtitle="2011"
        suggestionKey={0}
        title="A tvshow"
      />
      <SuggestionItem
        alreadyAdded={false}
        onPress={[Function]}
        poster="http://poster5.png"
        subtitle="2015"
        suggestionKey={1}
        title="A other tvshow"
      />
    </Styled(List)>
  </Styled(Content)>
</Styled(Container)>
`;

exports[`Rendering, with no suggestions should match 1`] = `
<Styled(Container)>
  <Styled(Content)>
    <Styled(Form)>
      <Styled(Item)
        fixedLabel={true}
      >
        <Styled(Label)>
          Tvshow’s name:
        </Styled(Label)>
        <Styled(Input)
          autoCapitalize="words"
          autoFocus={true}
          onChangeText={[Function]}
        />
      </Styled(Item)>
    </Styled(Form)>
    <Styled(List) />
  </Styled(Content)>
</Styled(Container)>
`;
